### ISSUE TITLE
Augmented Null-coalescing op and immutable hash tables

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#3612]
I don't think this should be an error:
```m2
i1 : h = new HashTable from { symbol cache => new CacheTable };

i2 : h.cache

o2 = CacheTable{}

o2 : CacheTable

i3 : try h.cache else h.cache = new CacheTable

o3 = CacheTable{}

o3 : CacheTable

i4 : h.cache ??= new CacheTable
stdio:4:8:(3): error: attempted to modify an immutable hash table
```

**Comment by mahrud** [comment]
Also, I'm confused about this outcome. I think it indicates that augmented null-coalescing operator acts like `=` rather than `:=`, but then why isn't `z` globally defined?
```m2
i1 : f = () -> ( y := null; y ??= 1; z ??= 2; );

i2 : peek User#"private dictionary"

o2 = GlobalDictionary{"f" => f}
                      "z" => z

i3 : f()

i4 : y

o4 = y

o4 : Symbol

i5 : z

o5 = z

o5 : Symbol
```

**Comment by d-torrance** [comment]
In general, `x ??= y` does the same thing as `x = x ?? y`.  Isn't this the expected behavior when `x` is immutable?

```m2
i1 : h = new HashTable from { symbol cache => new CacheTable };

i2 : h.cache =  h.cache ?? new CacheTable
stdio:2:8:(3): error: attempted to modify an immutable hash table
```

And in the second example, `z` isn't an error or null, so it's not assigned 2 -- it just stays as the symbol `z`.  It works if it's assigned as null:

```m2
i1 : f = () -> ( z = null; z ??= 2; );

i2 : f()

i3 : z

o3 = 2
```

**Comment by mahrud** [comment]
I see your point, but it would be very convenient if that syntax for making sure a hash table (mutable or not) has a key worked. For instance of `x ??= y` was instead interpreted as `x ?? (x = y)`.

**Comment by d-torrance** [comment]
What's the desired behavior?  If `h` is an immutable hash table doesn't have a `cache` key already, then we shouldn't be adding one later on.  If we want to do that, then we should have created it as a mutable hash table.

**Comment by mahrud** [comment]
Yes, if it's mutable and the key doesn't exist it should be an error, but if it does exist it should just return that. Say I have a function which accepts any hash table and caches something in it. Unfortunately some types already have a cache entry but some don't, so I would need to make sure it exists first and if it doesn't and the hash table is mutable, I can try to add it. That's the motivation.

**Comment by d-torrance** [comment]
Ok, I'm catching on!  I like this idea.  The `x ?? (x = y)` proposal also has the advantage of skipping the unnecessary step of re-assigning `x` to itself in the case when it's non-null, which has always bugged me.

**Comment by mahrud** [comment]
I imagine the downside is that this augmented operation would have to be dealt differently than the other augmented operations, but maybe that's worth it?

**Comment by d-torrance** [comment]
We were already dealing with it differently to begin with since we were suppressing any errors on the left-hand side.  I like this change -- great suggestion!

**Comment by d-torrance** [comment]
Fixed in #3614
